pytorch_lightning/callbacks/pruning.py

```diff
@@ -0,0 +1,282 @@
+                to specify dimension.
+
+            pruning_norm: if you are using ln_structured you need to specify norm.
+        """
+
+        self.use_global_unstructured = use_global_unstructured
+        self.parameters_to_prune = parameters_to_prune
+        self.prune_on_epoch_end = prune_on_epoch_end
+        self.prune_on_fit_end = prune_on_fit_end
+        self.use_lottery_ticket_hypothesis = use_lottery_ticket_hypothesis
+        self.parameter_names = parameter_names or self.PARAMETER_NAMES
+
+        for param_name in self.parameter_names:
+            if param_name not in self.PARAMETER_NAMES:
+                raise MisconfigurationException(
+                    f"The provided parameter_names {param_name} isn't in {self.PARAMETER_NAMES} "
+                )
+
+        if isinstance(pruning_fn, str):
+            if pruning_fn not in _PYTORCH_PRUNING_FUNCTIONS.keys():
+                raise MisconfigurationException(
+                    f"The provided pruning_fn {pruning_fn} isn't available with "
+                    f"PyTorch build-in {_PYTORCH_PRUNING_FUNCTIONS.keys()} "
+                )
+            if "unstructured" not in pruning_fn:
+                if pruning_dim is None:
+                    raise MisconfigurationException(
+                        "When requesting `structured` pruning, the `pruning_dim` should be provided."
+                    )
+                if pruning_fn == "ln_structured":
+                    if pruning_norm is None:
+                        raise MisconfigurationException(
+                            "When requesting `ln_structured` pruning, the `pruning_norm` should be provided."
+                        )
+
+                    pruning_fn = self.create_pruning_fn(pruning_fn, dim=pruning_dim, n=pruning_norm)
+            else:
+                pruning_fn = self.create_pruning_fn(pruning_fn)
+
+        self.pruning_fn = pruning_fn
+
+        if not (prune_on_epoch_end or prune_on_fit_end):
+            rank_zero_warn(
+                "The PruningCallback won't be triggered as not activate either on epoch_en or fit_end.", UserWarning)
+
+        self.make_pruning_permanent = make_pruning_permanent
+
+        if not isinstance(amount, (int, float, Callable)):
+            raise MisconfigurationException(
+                "amount should be provided and be either an int, a float or Callable function."
+            )
+
+        self.amount = amount
+
+    def create_pruning_fn(self, pruning_fn: str, *args, **kwargs):
+        if self.use_global_unstructured:
+            pruning_fn = _PYTORCH_PRUNING_METHOD[pruning_fn]
+            self._global_kwargs = kwargs
+            return pruning_fn
+        else:
+            return PruningCallback._wrap_pruning_fn(_PYTORCH_PRUNING_FUNCTIONS[pruning_fn], **kwargs)
+
+    @staticmethod
+    def _wrap_pruning_fn(pruning_fn, *args, **kwargs):
+        return lambda module, name, amount: pruning_fn(
```

Comment:
lambda functions are known to not work well with pickle, so this may give problems when running on multiple gpus

Reply:
You are right ! Let me see if I can something about it !

Reply:
Done !